charmhelpers/contrib/network/ip.py

```diff
@@ -493,14 +495,24 @@ def get_host_ip(hostname, fallback=None):
     if is_ip(hostname):
         return hostname
 
-    ip_addr = ns_query(hostname)
-    if not ip_addr:
-        try:
-            ip_addr = socket.gethostbyname(hostname)
-        except Exception:
-            log("Failed to resolve hostname '%s'" % (hostname),
-                level=WARNING)
-            return fallback
+    ip_addr = None
+    exc_value = None
+    try:
+        ip_addr = ns_query(hostname)
+    except Exception:
+        # If DNS lookup failed, try nsswitch (may check /etc/hosts).
+        # If that fails too, raise this exception/traceback instead.
+        (exc_type, exc_value, exc_traceback) = sys.exc_info()
+    finally:
+        if not ip_addr:  # Hit exception (non-NXDOMAIN) or no lookup/answer.
+            try:
+                ip_addr = socket.gethostbyname(hostname)
+            except Exception:
+                log("Failed to resolve hostname '%s'" % (hostname),
+                    level=WARNING)
+                if exc_value:
+                    reraise(exc_type, exc_value, exc_traceback)
+                return fallback
```

Comment:
I think this is a bit awkward for what it is doing.  The sequence is (if I understand the code):

1. ip_addr = ns_query(hostname)
2. ip_addr = socket.gethostbyname(hostname)

So this translates, for me, to:

```python
try:
    return ns_query(hostname)
except Exception as e:  # seems a bit too broad, maybe try to reign it in to a particular set of exceptions?
    try:
        return socket.gethostbyname(hostname)
    except Exception as f:
        log("failed to resolve hostname by either ns_query ({}) or socket.gethostbyname() ({})".format(str(e), str(f)))
        return fallback
```

Because, otherwise the new code in the patch does exactly what the old code does, but in a more complicated way.

Reply:
Uh oh, I've been caught trying to increase my LOC metrics. x)

So, it's not _exactly_ the same thing, because we now handle an exception in `ns_query()`.
That was the issue on T/X, which prevented even trying `gethostby{name,addr}()` for `/etc/hosts`.

The additional complication is, if we get to try `gethostby{name,addr}()` and it fails/raises exception,
we had discussed in the previous PR that it should reraise the exception generated in `ns_query()`
(to keep existing behavior in the library.)

If it's OK to change the behavior now (it wasn't back when we discussed this, if I understood it correctly)
and which exceptions/values are (not) raised/returned to the caller, then absolutely, the suggestion above is much simpler.

I can adjust the code according to what's the right direction now; just let me know.

Reply:
Okay, I think I've caught up.  (I may not have!)

So, the code is doing this?

```python
try:
    return ns_query(hostname)
except Exception as e:  # seems a bit too broad, maybe try to reign it in to a particular set of exceptions?
    t, v, tb = sys.exc_info()
    try:
        return socket.gethostbyname(hostname)
    except Exception as f:
        log("failed to resolve hostname by either ns_query ({}) or socket.gethostbyname() ({})".format(str(e), str(f)))
        if fallback:
            return fallback
        six.reraise(t, v, tb)
```

It would be nice not to have to use `six`, but the lib is *still* py27 compat.

Reply:
Hi @ajkavanagh 

Sorry, it's been a busy 2 weeks, and I had to review my own code and choices here. :)

So, the code is _not exactly_ doing that; let me try to clarify.
The complexity comes from sticking to no behavior changes. :/

- We need that `finally:` block, because in the original code `if not ip_addr:` runs inconditionally, not just in case of unhandled exception in `ns_query()` (it can return `None` for `NXDOMAIN`).

- On the `fallback` check/return before `reraise`: good point.
If `fallback` is set, we should not `reraise`, but use that instead.

- Regarding `six`, I've used it to keep the backtraces in py2 consistent or at least useful; py3 has exception chaining, but IIRC in py2 (which doesn't) if we just `raise` the original exception, we lose the backtrace altogether; bad enough for debugging.
And then while I was trying to use a python exception thing one way or another, it would be a syntax failure or other problem in either python version.
So I eventually gave up on syntatic sugar and chose `six` (with this verbose `t, c, tb = exc_info()` approach.)

Hope this helps!